### PR TITLE
WIP: Adding training script

### PR DESCRIPTION
**Pull request by isamu-isozaki**
This is a work in progress pr. For now, the plan is

- [x] Add BaseTrainer
- [x] Abstract away parts of the VQGANVAETrainer to the BaseTrainer
- [x] Add BaseMaskGitTrainer
- [x ] Add SuperresMaskGitTrainer
- [ ] Make training script for vae in train_muse_vae.py
- [ ] Make training script for  mask git trainer in train_maskgit.py
- [ ] Add inference script in infer_muse.py
- [ ] Add text and image dataset integrated with huggingface datasets
- [x] Tokenize text within the dataset
- [ ] Add support for colossalai

There's quite a bit of detail missing here but overall this is the idea. 
@lucidrains Will ping you when I'm done! A

### CODE REVIEW

**Comment by lucidrains** [comment]
@isamu-isozaki Boss.

thank you!

**Comment by isamu-isozaki** [comment]
@ZeroCool940711 fixed some of the bugs so nearly done with a basic scripts!

**Comment by isamu-isozaki** [comment]
vae training script seems to be working with wandb/tensorboard integrated! Next trying out mask git

**Comment by isamu-isozaki** [comment]
Got training for both base maskgit and vae working!

**Comment by isamu-isozaki** [comment]
I'll test a bit more and then move on to the other tasks above.

**Comment by lucidrains** [comment]
> Got training for both base maskgit and vae working!

wow! :100: you should get in touch with huggingface, specifically @patrickvonplaten

they are looking at replicating this!

**Comment by isamu-isozaki** [comment]
@lucidrains haha thank you! Sent him a message. And thanks for the kind words. Wouldn't have been able to do it without your great work!

**Comment by isamu-isozaki** [comment]
Currently @ZeroCool940711 is testing if it properly trains in the long term

**Comment by isamu-isozaki** [comment]
@wnakano Added some code to support loading text from directories+adding support for multiple descriptions+fixing the tokenizer. Training with text+image pairs from directories should now be way more managable

**Comment by isamu-isozaki** [comment]
He's next working on trying loading maskgit weights from the google repo which will be pretty huge!

**Comment by isamu-isozaki** [comment]
Added some qol such as saving a folder of image text pairs in a huggingface dataset. Now, @ZeroCool940711 is training the base maskgit on resolution 384x384 on his INE dataset and we'll later test on cifar10 and mnist as well!

**Comment by isamu-isozaki** [comment]
@ZeroCool940711 added resume paths for both maskgit and vae so you can resume training!

**Comment by isamu-isozaki** [comment]
This is a wandb run on base maskgit on mnist. The prompt is 4 but we accidentally enabled flipping image for augmentation. Check out the wandb run [here](https://wandb.ai/tukirito/muse_maskgit/runs/7xsivy6s?workspace=user-isamu) it was trained by @ZeroCool940711 on 8gb of gpu and all the train parameters are there. He's also uploading checkpoints to
[here](https://huggingface.co/Sygil/Sygil-Muse)

**Comment by isamu-isozaki** [comment]
Main params: batch size 400, image size 32, cosine_with_restarts, warm up steps 1000

**Comment by isamu-isozaki** [comment]
We got 7 in less than 1k steps [here](https://wandb.ai/tukirito/muse_maskgit/runs/bkb2e7yg?workspace=user-isamu)

**Comment by ZeroCool940711** [comment]
It took less than 1K steps for it to be able to generate the next number correctly, it was a 4 before and now a 7, the first sample was more like a 1 than a 7 but after around 600 steps more of training the samples started to look more and more like 7s.

**Comment by isamu-isozaki** [comment]
We plan to next move onto more complicated datasets like cifar10 and hopefully test out super res training too

**Comment by lucidrains** [comment]
wow, you kids are incredible. and fast!

maybe you'll be part of the next generation of AI engineers at [Cerebral Valley](https://sfstandard.com/technology/what-is-cerebral-valley-san-franciscos-nerdiest-new-neighborhood/) 😂

**Comment by isamu-isozaki** [comment]
@lucidrains not as fast as you! And haha thanks I'll take that as a compliment.

**Comment by ZeroCool940711** [comment]
I feel out of place for some reason. Most of you guys with PHDs and others going to college, then somehow there is also this guy (me) who didn't even finish high school and don't even consider himself a programmer working on this stuff with you guys :V
well, I'm having fun with this so that's probably what counts.

**Comment by patrickvonplaten** [comment]
Wow amazing work here @isamu-isozaki! cc'ing @patil-suraj here as well

**Comment by isamu-isozaki** [comment]
Update. @wnakano added code for loading pretrained vaes from [taming transformers](https://github.com/CompVis/taming-transformers) and [latent diffusion](https://github.com/CompVis/latent-diffusion). So less training is needed most likely!

For my end, I'm working with @dsbuddy, who I worked on the medsegdiff training script with, and I think we are close to finishing training the vqvae for cifar10 and will move on to base maskgit training on cifar10 tomorrow.

**Comment by isamu-isozaki** [comment]
@patrickvonplaten oh awesome seeing you here!

**Comment by isamu-isozaki** [comment]
Hi! Brief update. Laion seems pretty interested in first improving the vqgan so we are trying that out now! The main improvements we are thinking for now are
1. Add an option for a vit vqgan
2. Add in wavelet transforms similar to simple diffusion mainly from crowsonkb's scripts so that high frequency noise is removed
3. Add option for other timm models in place of the vgg discriminator

We are still reading research papers on vqgans so if anyone has any recommendations for papers let us know!

**Comment by isamu-isozaki** [comment]
@lucidrains Sorry forgot to do an update here for close to 2 months haha.
I mainly moved to huggingface/open-muse [here](https://github.com/huggingface/open-muse). And we are doing some pretty exciting stuff! We already got some pretty good results on Imagenet so we are mainly scaling up+combining insights from paella. Once that's done happy to finish this pr!
